go/store/nbs/table_index.go

```diff
@@ -0,0 +1,506 @@
+		prevOff = 0
+	} else {
+		prevOff = ti.offsetAt(ord - 1)
+	}
+	ordOff := ti.offsetAt(ord)
+	length := uint32(ordOff - prevOff)
+	return indexResult{
+		o: prevOff,
+		l: length,
+	}
+}
+
+func (ti onHeapTableIndex) Lookup(h *addr) (indexEntry, bool, error) {
+	ord, err := ti.lookupOrdinal(h)
+	if err != nil {
+		return indexResult{}, false, err
+	}
+	if ord == ti.chunkCount {
+		return indexResult{}, false, nil
+	}
+	return ti.getIndexEntry(ord), true, nil
+}
+
+// lookupOrdinal returns the ordinal of |h| if present. Returns |ti.chunkCount|
+// if absent.
+func (ti onHeapTableIndex) lookupOrdinal(h *addr) (uint32, error) {
+	prefix := h.Prefix()
+
+	for idx := ti.prefixIdx(prefix); idx < ti.chunkCount && ti.prefixAt(idx) == prefix; idx++ {
+		m, err := ti.EntrySuffixMatches(idx, h)
+		if err != nil {
+			return ti.chunkCount, err
+		}
+		if m {
+			return ti.ordinalAt(idx), nil
+		}
+	}
+
+	return ti.chunkCount, nil
+}
+
+// prefixIdx returns the first position in |tr.prefixes| whose value ==
+// |prefix|. Returns |tr.chunkCount| if absent
+func (ti onHeapTableIndex) prefixIdx(prefix uint64) (idx uint32) {
+	// NOTE: The golang impl of sort.Search is basically inlined here. This method can be called in
+	// an extremely tight loop and inlining the code was a significant perf improvement.
+	idx, j := 0, ti.chunkCount
+	for idx < j {
+		h := idx + (j-idx)/2 // avoid overflow when computing h
+		// i ≤ h < j
+		if ti.prefixAt(h) < prefix {
+			idx = h + 1 // preserves f(i-1) == false
+		} else {
+			j = h // preserves f(j) == true
+		}
+	}
+
+	return
+}
+
+func (ti onHeapTableIndex) tupleAt(idx uint32) (prefix uint64, ord uint32) {
+	off := int64(prefixTupleSize * idx)
+	b := ti.tupleB[off : off+prefixTupleSize]
+
+	prefix = binary.BigEndian.Uint64(b[:])
+	ord = binary.BigEndian.Uint32(b[addrPrefixSize:])
+	return prefix, ord
+}
+
+func (ti onHeapTableIndex) prefixAt(idx uint32) uint64 {
+	off := int64(prefixTupleSize * idx)
+	b := ti.tupleB[off : off+addrPrefixSize]
+	return binary.BigEndian.Uint64(b)
+}
+
+func (ti onHeapTableIndex) ordinalAt(idx uint32) uint32 {
+	off := int64(prefixTupleSize*idx) + addrPrefixSize
+	b := ti.tupleB[off : off+ordinalSize]
+	return binary.BigEndian.Uint32(b)
+}
+
+func (ti onHeapTableIndex) offsetAt(ord uint32) uint64 {
+	off := int64(offsetSize * ord)
+	b := ti.offsetB[off : off+offsetSize]
+	return binary.BigEndian.Uint64(b)
+}
+
+func (ti onHeapTableIndex) Ordinals() ([]uint32, error) {
+	o := make([]uint32, ti.chunkCount)
+	for i, off := uint32(0), 0; i < ti.chunkCount; i, off = i+1, off+prefixTupleSize {
+		b := ti.tupleB[off+addrPrefixSize : off+prefixTupleSize]
+		o[i] = binary.BigEndian.Uint32(b)
+	}
+	return o, nil
+}
+
+func (ti onHeapTableIndex) Prefixes() ([]uint64, error) {
+	p := make([]uint64, ti.chunkCount)
+	for i, off := uint32(0), 0; i < ti.chunkCount; i, off = i+1, off+prefixTupleSize {
+		b := ti.tupleB[off : off+addrPrefixSize]
+		p[i] = binary.BigEndian.Uint64(b)
+	}
+	return p, nil
+}
```

Comment:
Where are these used? Conjoin? They seem less-than-ideal for an API from a memory consumption point of view...

Reply:
Correct. Conjoin. Agreed, although I haven't thought too deeply about what the tableIndex interface should look like. It depends on how we reimplement conjoin without loading indexes into memory.